### PR TITLE
Driver for XPT2046 touch controller

### PR DESCRIPTION
**Pull request by FrankX0**
Fixes #1242 .

- [X] This PR is for the `dev` branch rather than for `master`.
- [X] This PR is compliant with the [other contributing guidelines](https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md) as well (if not, please describe why).
- [X] I have thoroughly tested my contribution.
- [X] The code changes are reflected in the documentation at `docs/en/*`.

This driver is based on the work of @Starofall with some changes and allowing the touch position to be retrieved using a callback from a gpio interrupt (see example in the documentation).
Main issue with this implementation is the requirement to disable and clear the interrupts which are being generated while reading the touch information from the touch controller. I specifically request feedback whether my implementation is acceptable or proposals for improvements.

### CODE REVIEW

**Comment by marcelstoer** [comment]
Looking forward to this.
Did you study the feedback in #1242? We agreed that a single module and specific controllers which derive from it would ideally be the way to go.

**Comment by FrankX0** [comment]
I read the feedback in #1242. My current skills are insufficient to start such a module.
Also not sure if it is worth the investment, as there are not so many touch controllers for small LCDs.
Saying that, I am working on a driver for a capacitive touch controller. I would be interested to help out defining the interface of such a single touch module, covering both resistive and capacitive controllers.

**Comment by marcelstoer** [comment]
@all specifically @devsaurus how shall we proceed with this? Do we insist on "single module and specific controllers/drivers"? If so, then we should close this PR.

**Comment by devsaurus** [comment]
Don't want to be too strict, to be honest. Guess we could consider this PR.
Though the topic would be raised again if another touchscreen driver comes around.

**Comment by FrankX0** [comment]
I agree. The number of touch controllers will be limited anyway.
In my view this one will be quite useful for the community (at least for me it does).

File: app/modules/xpt2046.c [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+// Hardware specific values
+static const uint16_t CAL_MARGIN = 0; // Set to 0: up to the application
+static const uint8_t CTRL_LO_DFR = 0b0011;
+static const uint8_t CTRL_LO_SER = 0b0100;
+static const uint8_t CTRL_HI_X = 0b1001  << 4;
+static const uint8_t CTRL_HI_Y = 0b1101  << 4;
+static const uint16_t ADC_MAX = 0x0fff;  // 12 bits
+
+// Runtime variables
+uint16_t _width, _height;
```
Comment (devsaurus) [inline app/modules/xpt2046.c]: This and below: why not declared `static`?

File: app/modules/xpt2046.c [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+  if ( da <= db && da <= dc ) reta = (x + y) >> 1;
+  else if ( db <= da && db <= dc ) reta = (x + z) >> 1;
+  else reta = (y + z) >> 1;   
+
+  return reta;
+}
+
+// Checks if the irq_pin is down
+static int isTouching() { 
+	return (platform_gpio_read(_irq_pin) == 0);
```
Comment (devsaurus) [inline app/modules/xpt2046.c]: Please use consistent indentation.

File: docs/en/modules/xpt2046.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
+# XPT2046 Module
+| Since  | Origin / Contributor  | Maintainer  | Source  |
+| :----- | :-------------------- | :---------- | :------ |
+| 2016-04-14| [Starofall](https://github.com/Starofall) | [Frank Exoo](https://github.com/FrankX0) | [xpt2046.c](../../../app/modules/xpt2046.c)|
+
+XPT2046 is a touch controller used by several cheap displays - often in combination with the ILI9341 display controller.
+The module is build based on the libarys of [spapadim](https://github.com/spapadim/XPT2046/) and  [PaulStoffregen](https://github.com/PaulStoffregen/XPT2046_Touchscreen).
+
+
+## xpt2046.init(cs_pin, irq_pin, height, width)
```
Comment (devsaurus) [inline docs/en/modules/xpt2046.md]: Here and below: Keep API doc consistent and don't list the parameters in the section title.

Comment (marcelstoer) [inline docs/en/modules/xpt2046.md]: ...is buil**t**..., there's a superfluous space before the second link

File: docs/en/modules/xpt2046.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+Initiates the XPT2046 module to read touch values from the display. It is required to call spi.setup() before calling xpt2046.init (see example).
+As the ucg lib also requires spi.setup() to be called before it is important to only call it once in total and to activate spi.FULLDUPLEX.
+The `clock_div` used in spi.setup() should be 16 or higher, as lower values might produces inaccurate results.
+
```
Comment (marcelstoer) [inline docs/en/modules/xpt2046.md]: Please add links to the function references.

File: docs/en/modules/xpt2046.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+#### Parameters
+- `cs_pin` GPIO pin for cs
+- `irq_pin` GPIO pin for irq
+- `height` display height in pixel
+- `width` display width in pixel
+
+#### Returns
+`nil`
+
```
Comment (marcelstoer) [inline docs/en/modules/xpt2046.md]: You're missing an "Example" heading here.

**Comment by marcelstoer** [comment]
Thanks Frank, in case you happen to have a complete example script laying around you could add that to the `/lua_examples` folder in the repo.

**Comment by marcelstoer** [comment]
Superseded by #1848 -> closing

**Pull request closed without merging**